### ISSUE TITLE
Banish the evil null with None in GDScript for object based checks

### ISSUE THREAD

**Issue by chucklepie** [godotengine/godot-proposals#2339]
### Describe the problem or limitation you are having in your project
Despite the guards put in place, nulls can cause segmentation faults at runtime for GDScript, where workarounds have to be carefully put in place, for example using weak references or the 3.1 addition 'is_instance_valid' method.

But it remains that you can still code for 
```
if mynode == null
```
and get an error/crash because nobody really thinks that a node might be in an intermediate stage of deletion or wants to type a longer conditional check to check if null or it is deleted.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Python removed null completely from the language and for good reason. Using weak references is long-winded and far too complex for your average gdscript, and typing 'if is_valid_instance(node)' every time you want to check a node exists or not is cumbersome.

My feature would be similar to python None

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Simply

```
if mynode == None
```

or
```
if mynode is None
```

Whichever is better suited. Would simply translate to an internal null check and a check whether the node exists or not, i.e. 'is_instance_valid'

This would fix so much code that uses null checks, and help banish a large portion where nodes are null checked.

**Comment by Calinou** [comment]
> This would fix so much code that uses null checks

People would still have to modify their existing scripts and make sure not to compare nodes against `null` in the future :slightly_smiling_face:

**Comment by chucklepie** [comment]
Yes, that's what's been happening for the past 10 years in C#. You improve and change :)

Null is still there, just make a 'safer' version of it like Python that doesn't fall foul of 'deleted object', etc.

**Comment by pycbouh** [comment]
Well, you mention yourself, that there is already a safer version. If you would need to teach people a new syntax, wouldn't it be easier to teach them the method call that is already a part of the API, and avoid additional questions about how `None` compares to `null`?

**Comment by chucklepie** [comment]
@pycbouh GDscript has 'is' which is exactly the same problem of portability...

I'm just suggesting a nice fluent way of avoiding crashes due to not always knowing if your objects are deleted but not null instead of having to wrap your code around an api call when a language construct is far more intuitive....

**Comment by pycbouh** [comment]
Many languages have `is` 🙂 

To that end why not make `null` checks themselves safer for `Node`s? We fully control the language environment, what would stop us from doing the same logic without a new keyword/special type?

> I refer back to my comment on Python not even having null.

That's not really useful. Python may have or omit many things, but it is not relevant. We are not bound by the same constraints as Python maintainers.

**Comment by chucklepie** [comment]
Nothing, I just presumed from a performance point of view the C++ null operator was much simpler than a 'is null or deleted' check?

This all comes, btw, from a 6 hour debug trying to figure out why I was getting a segmentation fault and it was calling a property on a deleted object ;-)
